### PR TITLE
merging & making grid boundaries

### PR DESCRIPTION
**Pull request by ghost**
Hello! Can you check my code?

### CODE REVIEW

File: constants.h [hunk 1]
```diff
@@ -29,7 +29,7 @@ struct Vector{ // TODO: is this needed for velocity?
         this->y += rhs.y;
         return *this;
     }
-    friend Vector operator+(Vector<T> lhs, const Vector<T>& rhs ){
+    friend Vector operator+(Vector<T> lhs, const Vector<T>& rhs){
```
Comment (Randl) [inline constants.h]: please use code formatter (preferably as git hook), e.g., clang-format
https://www.jetbrains.com/help/clion/clangformat-as-alternative-formatter.html

File: point.h [hunk 2]
```diff
@@ -13,47 +13,45 @@
...
     bool exist;
 
     Vector<double> v;
-    void eq(); // TODO: check
+    void eq();
     void col(); // TODO: implement
     void macro(); // TODO:check
     Point();
-    Point(std::array<double,5>);// TODO: initials
+    Point(std::array<double, 4>);// TODO: initials
```
Comment (Randl) [inline point.h]: don't use magic constant. what is 4?

File: point.h [hunk 3]
```diff
@@ -13,47 +13,45 @@
...
 
 };
 class Grid{
     Grid(std::vector<std::pair<int, int>>);
     std::vector<std::vector<Point>> points;
     bool is_possible(int, int,int);  // TODO: is needed?
     void mirroring(int,int,int);  // TODO: boundaries
+    void boundaries(Point **desc_grid);
```
Comment (Randl) [inline point.h]: don't use raw pointers, I assume it is a 2D array? Use `std::vector<std::vector<Point>>` then

File: point.h [hunk 4]
```diff
@@ -145,8 +143,31 @@ void Grid::transfer(int x, int y) {
         if (is_possible(x,y,k)) points[x+e[k].x][y+e[k].y].f_temp[k] = points[x][y].f[k];
     }
 }
-
-
+void Grid::boundaries(Point **grid){
```
Comment (Randl) [inline point.h]: Document any added function, use doxygen:
https://www.jetbrains.com/help/clion/creating-and-viewing-doxygen-documentation.html

File: point.h [hunk 5]
```diff
@@ -145,8 +143,31 @@ void Grid::transfer(int x, int y) {
...
+    bound_grid = new bool*[sizeof(grid)];
```
Comment (Randl) [inline point.h]: don't use C arrays, use vectors

File: point.h [hunk 6]
```diff
@@ -145,8 +143,31 @@ void Grid::transfer(int x, int y) {
...
+    for (unsigned int i = 0; i < sizeof(grid); i++)
+        bound_grid[i] = new bool[sizeof(*grid)];
+    for(int i = 0; i < sizeof(grid); ++i){
+        for(int j = 0; j < sizeof(*grid); ++j){
+            if(grid[i][j].exist){
+                if (i==0 || j==0 || i==sizeof(grid) || j==sizeof(*grid)){
+                    bound_grid[i][j] = 1;// Если точка на границе сетки, то она -- граница
```
Comment (Randl) [inline point.h]: be consistent with comment language, better use English

**Review by Randl — CHANGES_REQUESTED** [review]
Replace C arrays with vectors everywhere

File: point.h (hunk 2 above)
Comment (EzAccount) [reply point.h]: Should be 5: rho, T, P, v_x, v_y

File: point.h [hunk 7]
```diff
@@ -13,47 +13,45 @@
...
         f[k] = f_temp[k] - 1./tau * (f_temp[k] -  f_eq[k]);
     }
 }
 void Point::eq() {
     double c = T;
-    for (size_t k=0; k<Q; ++k){
+    for (size_t k = 0; k < Q; ++k){
         double sc = e[k] * v;
-        f_eq[k] = rho * w[k] * (1. + 3 * sc/c + 4.5 * sc * sc / c / c -  v * v * 1.5/ c);
+        f_eq[k] = rho * w[k] * (1. + 3 * sc/c + 4.5 * sc * sc / c / c -  v * v * 1.5/ c / c);
```
Comment (EzAccount) [inline point.h]: check coefficients in series expansion based on dimenshionalless form of the equation (either we use v^2=2RT or 3RT)

File: point.h [hunk 8]
```diff
@@ -85,14 +83,14 @@ Point::Point() {
         f_temp[k] = f_eq[k];
     }
 }
-Point::Point(std::vector<double> init){
+Point::Point(std::array<double, 4> init){
```
Comment (EzAccount) [inline point.h]: +pressure

File: point.h [hunk 9]
```diff
@@ -145,8 +143,31 @@ void Grid::transfer(int x, int y) {
...
-
-
+void Grid::boundaries(Point **grid){
+    bound_grid = new bool*[sizeof(grid)];
+    for (unsigned int i = 0; i < sizeof(grid); i++)
+        bound_grid[i] = new bool[sizeof(*grid)];
+    for(int i = 0; i < sizeof(grid); ++i){
+        for(int j = 0; j < sizeof(*grid); ++j){
+            if(grid[i][j].exist){
+                if (i==0 || j==0 || i==sizeof(grid) || j==sizeof(*grid)){
```
Comment (EzAccount) [inline point.h]: PEP8 spaces

File: point.h [hunk 10]
```diff
@@ -145,8 +143,31 @@ void Grid::transfer(int x, int y) {
...
+                    bound_grid[i][j] = 1;// Если точка на границе сетки, то она -- граница
+                }
+                else {
+                    int count = 0;
+                    for (int a = -1; a < 1; ++a) {
```
Comment (EzAccount) [inline point.h]: Probably a <= 1 expected

File: point.h (hunk 5 above)
Comment (EzAccount) [reply point.h]: local variable?

File: point.h [hunk 11]
```diff
@@ -13,47 +13,45 @@
...
+    void boundaries(Point **desc_grid);
     void transfer(int, int); // TODO: standalone function
     void at(int, int); // TODO: is this needed?
     void eval();
 public:
     std::array<double,5> macro_at(size_t, size_t); // TODO: result output
-public:
     Point **desc_grid;
-
+    bool **bound_grid;
```
Comment (EzAccount) [inline point.h]: name is not clear

File: point.h [hunk 12]
```diff
@@ -13,47 +13,45 @@
...
     bool is_possible(int, int,int);  // TODO: is needed?
     void mirroring(int,int,int);  // TODO: boundaries
+    void boundaries(Point **desc_grid);
     void transfer(int, int); // TODO: standalone function
     void at(int, int); // TODO: is this needed?
     void eval();
 public:
     std::array<double,5> macro_at(size_t, size_t); // TODO: result output
-public:
     Point **desc_grid;
```
Comment (EzAccount) [inline point.h]: name is not clear

**Review by EzAccount — CHANGES_REQUESTED** [review]
Mark all as resolved when fixed and apply for new review

File: point.h (hunk 2 above)
Comment (Randl) [reply point.h]: or, for constructor pass everything as parameters

**Comment by EzAccount** [comment]
Minor thing: don't use bad words or russian in comments and/or on github. This project will be open to anyone, be professional

File: main.cpp [hunk 13]
```diff
@@ -25,7 +25,6 @@ double*  vec_mul(double *a, double b) // vector * scalar
     return a;
 }*/ //Do i need this func?
```
Comment (Randl) [inline main.cpp]: Don't leave commented out code, you have VCS for that

File: point.h [hunk 14]
```diff
@@ -4,151 +4,144 @@
...
-    rho = init[0];
-    v.x = init[1];
-    v.y = init[2];
-    eq();
-    for (int k=0; k<Q; ++k){
-        f[k] = f_eq[k];
-        f_temp[k] = f_eq[k];
-    }
+
+Point::Point(double rho, double T, Vector<double> v, double P)
```
Comment (Randl) [inline point.h:64]: Just give default values to all the parameters to not duplicate code with default constructor

File: point.h [hunk 15]
```diff
@@ -4,151 +4,144 @@
...
-                    flag = 1;
-                    break;
-                }
-            }
-            if (flag==0){
-                desc_grid[i][j].exist = 0;
-            } else flag = 0;
+Grid::Grid(std::vector<std::pair<int, int>> indata) {
+  int xmax = 0, ymax = 0, xmin = 0, ymin = 0;
+  xmax = indata[0].first;
```
Comment (Randl) [inline point.h]: why init with 0 and rewrite immediately? Just init in previous line

File: point.h [hunk 16]
```diff
@@ -4,151 +4,144 @@
...
+  xmin = indata[0].first;
+  ymax = indata[0].second;
+  ymin = indata[0].second;
+  for (int i = 0; i < indata.size(); ++i) {
+    xmax = std::max_element(xmax, indata[i].first);
```
Comment (Randl) [inline point.h]: Check the documentation for `std::max_element`. You are not supposed to have any loop here

File: point.h [hunk 17]
```diff
@@ -4,151 +4,144 @@
...
-}
 void Grid::transfer(int x, int y) {
-    for (size_t k=0; k<Q; ++k){
-        if (is_possible(x,y,k)) points[x+e[k].x][y+e[k].y].f_temp[k] = points[x][y].f[k];
+  for (size_t k = 0; k < Q; ++k) {
+    if (grid[x][y].exist)
+      grid[x + e[k].x][y + e[k].y].f_temp[k] = points[x][y].f[k];
+  }
+}
+void Grid::boundaries() {
```
Comment (Randl) [inline point.h:125]: documentation should be here, not in declaration.

File: constants.h [hunk 18]
```diff
@@ -4,57 +4,55 @@
...
+  }
+  friend Vector operator*(T scalar, const Vector<T> &rhs) {
+    return Vector(scalar * rhs.x, scalar * rhs.y);
+  }
 };
 
-
 constexpr std::size_t Q = 9;
-double Pr = 2./3;
+double Pr = 2. / 3;
```
Comment (Randl) [inline constants.h:43]: better add description for each constant

File: constants.h [hunk 19]
```diff
@@ -4,57 +4,55 @@
...
-    }
-    T operator*(const Vector<T>&rhs) const{
+template <typename T> struct Vector {
+  T x;
+  T y;
+  Vector() {
+    x = 0;
+    y = 0;
+  }
+  Vector(T a, T b) {
```
Comment (Randl) [inline constants.h]: Better use initializer list like in `Point` constructor:
```Vector(T x=0, T y=0) : x(x), y(y) {}```

File: point.h [hunk 20]
```diff
@@ -4,151 +4,144 @@
...
+#include <array>
 #include <iostream>
 #include <vector>
-#include <array>
-#include <algorithm>
-#include "constants.h"
-
 
-class Point{
+class Point {
```
Comment (Randl) [inline point.h:13]: you might want make point templated on data type (double) in case you'll need other data type (say, float).

File: point.h [hunk 21]
```diff
@@ -4,151 +4,144 @@
...
-    void at(int, int); // TODO: is this needed?
-    void eval();
-public:
-    std::array<double,5> macro_at(size_t, size_t); // TODO: result output
-public:
-    Point **desc_grid;
-
+class Grid {
+  Grid(std::vector<std::pair<int, int>>);
+  /***
```
Comment (Randl) [inline point.h]: this belongs to implementation

**Review by Randl — CHANGES_REQUESTED** [review]
Most of problems are minor, but you have algorithmic problem in maximum detection, so at least fix that.

File: point.h [hunk 22]
```diff
@@ -4,172 +4,144 @@
...
-    }
-    return false;
-}
 void Grid::transfer(int x, int y) {
-    for (size_t k=0; k<Q; ++k){
-        if (is_possible(x,y,k)) points[x+e[k].x][y+e[k].y].f_temp[k] = points[x][y].f[k];
-    }
+  for (size_t k = 0; k < Q; ++k) {
+    if (grid[x][y].exist)
+      grid[x + e[k].x][y + e[k].y].f_temp[k] = points[x][y].f[k];
```
Comment (EzAccount) [inline point.h]: what if grid[x][y] is boundary?

**Review by Randl — APPROVED** [review]

File: point.h [hunk 23]
```diff
@@ -105,17 +105,17 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
 }
 
 void Grid::transfer(int x, int y) {
-  if (grid[x][y].bound) {
-    for (size_t k = 0; k < Q; ++k) {
+  for (size_t k = 0; k < Q; ++k) {
+    if (grid[x][y].bound) {
       if (grid[x + e[k].x][y].bound && !grid[x][y + e[k].y].bound) {
         e[k].y = -e[k].y;
```
Comment (EzAccount) [inline point.h:111]: углы?

File: point.h [hunk 24]
```diff
@@ -105,17 +105,17 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
 }
 
 void Grid::transfer(int x, int y) {
```
Comment (EzAccount) [inline point.h:107]: any tests on transfer method?

File: point.h [hunk 25]
```diff
@@ -105,17 +105,17 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
...
-  if (grid[x][y].bound) {
-    for (size_t k = 0; k < Q; ++k) {
+  for (size_t k = 0; k < Q; ++k) {
+    if (grid[x][y].bound) {
       if (grid[x + e[k].x][y].bound && !grid[x][y + e[k].y].bound) {
```
Comment (EzAccount) [inline point.h:110]: what about corners?

**Review by EzAccount — CHANGES_REQUESTED** [review]

**Pull request merged**